src/osd/OSD.cc

```diff
@@ -4304,11 +4304,12 @@ void OSD::build_initial_pg_history(
   OSDMapRef lastmap = service.get_map(created);
   int up_primary, acting_primary;
   vector<int> up, acting;
-  lastmap->pg_to_up_acting_osds(
-    pgid.pgid, &up, &up_primary, &acting, &acting_primary);
 
   ostringstream debug;
   for (epoch_t e = created + 1; e <= osdmap->get_epoch(); ++e) {
+    lastmap->pg_to_up_acting_osds(
+      pgid.pgid, &up, &up_primary, &acting, &acting_primary);
+
```

Comment:
Good catch.  I think a better fix, though, is to add at the end of the loop
<pre>
up = new_up;
acting = new_acting;
up_primary = new_up_primary;
acting_primary = new_acting_primary;
</pre>
and avoid calculating each mapping twice